File: scala/p2.scala
object p2 { 
  def fib(a: Int, b: Int, lst: List[Int]) : List[Int] = if (a+b <= 4000000) fib(b, a+b, (a+b)::lst) else lst  
  def a() = fib(0,1,List(0)).reduceLeft[Int] { (t,n) => if (n % 2 == 0) t+n else t }

  def main(args: Array[String])
  {
    println(a())
  }
}

File: haskell/p2.hs
--fib = 0 : 1 : zipWith (+) fib (tail fib)
--main = do print(take 3 fib) 
     -- print(sum [i | i <- take 15 fib, even i])
f1 = [0,1,2,3,4,5]
f2 = 0 : zipWith (+) f2 (tail f1)
main = do print (f2)